### ISSUE TITLE
Running Kinbot Code with Neural Network Potential

### ISSUE THREAD

**Issue by IliasChair** [zadorlab/KinBot#79]
Hello,
I'm interested in using Kinbot with a Neural Network Potential (NNP). The codebase mentions NN PES capabilities in several places, and the presence of the 'nn_pes.py' script seems to confirm it.

However, I encountered an issue while trying to run the code. Specifically, I couldn't import the 'util' package imported in the nn_pes.py script, which appears to contain functions for using the NN PES. I haven't been able to locate this package or any references to the functions it contains within the rest of the codebase.

Any guidance on setting up and running Kinbot with an NNP would be greatly appreciated.

**Comment by cmartia** [comment]
Hi Ilias,
Thanks for trying out KinBot. Do you have your own NN potential that you want to use with KinBot? Or are you trying to use the one we developed for the C5H5 system? 

In the first case, KinBot should work with any NN potential that is implemented as an ASE calculator. In principle you just need to clone this repo, and substitute the file `KinBot/kinbot/ase_modules/calculators/nn_pes.py` with the file containing your NN calculator. It's possible some very minor tweaking needs to be made for generalization. In that case I would be happy to assist you. 

In the second case, you need to install the pykinml package (https://pypi.org/project/pykinml/) as the NN per se is independent of KinBot. I'll be also happy to assist you if you have any problems.

Carles

**Comment by IliasChair** [comment]
Hi Carles, 
thank you very much for the swift reply. I plan to use the code provided by [FAIR-Chem/OPC](https://github.com/FAIR-Chem/fairchem/tree/main). They provide an interface with ASE to calculate forces and energies.

However, I think getting this to work is not as simple as just passing the ASE calculator to Kinbot. From what I can tell, the 'nn_pes.py' script implements and expects functionality beyond just calculating energy and forces. If I were to reimplement this, I would need to know which functionality this script must provide. 

Furthermore there are mentions to 'Nn_surr' and 'nn_pes' in the codebase. If 'surr' is short for 'surrogate' I dont understand what the difference between them is, as in literature they are usually interchangable.

Much obliged
Ilias

**Comment by cmartia** [comment]
The file nn_pes.py as it is now, implements a NN potential for the C5H5 system which depends on the pykinml package. However, if you substitute that file with your own it will depend on what the contents of that file. The only requirements this file needs to have is being an ase calculator and it is true that the name of the Calculator class should be Nn_surr. Maybe I am forgetting something but I think there should be no conflicts. Let me know otherwise.

**Comment by cmartia** [comment]
One more thing. In the KinBot input file. you should specify `"qc" : "gauss",`

**Comment by cmartia** [comment]
Here there is an example input file. You should substitute the last block with the parameters relevant to your computing center. 

`{
"title": "c5h5",
"structure": [
        "C", 0.3738239349609347, 0.14447877568430476, -0.14043504101665275,
        "C", 1.632078036032578, -0.19526893054975314, 0.39150489887490575,
        "C", 0.4568145219646841, 1.4751014718969684, -0.6817029011188551,
        "C", 2.4919897755620743, 0.9007317691780267, 0.1892700059979383,
        "C", 1.7521659157918774, 1.937585643689521, -0.4800794287621648,
        "H", -0.5086826145412491, -0.48082442668309394, -0.14566009844899153,
        "H", 1.891471197821385, -1.1302331237235264, 0.8696085186941203,
        "H", 2.145237308341487, 2.9024097987011204, -0.7701132596713715,
        "H", 3.5311167196581406, 0.9615218780753692, 0.4831555463511647,
        "H", -0.3525122380930552, 2.0106343480310085, -1.158889568906007
      ],
"charge": 0,
"mult": 2,

"pes": 0,
"barrier_threshold": 105.0,
"high_level": 0,
"L3_calc": 0,
"verbose": 0,

"use_sella": true,
"sella_kwargs": {"delta0": 1e-2},
"qc": "nn_pes",

"scan_method": "am1",
"scan_basis": "",
"method": "",
"basis": "",
"high_level_method": "",
"high_level_basis": "",

"conformer_search": 1,
"max_dihed": 4,
"random_conf": 300,

"rotor_scan": 1,
"nrotation": 24,

"irc_maxpoints": 30,
"irc_stepsize":10,

"me": 1,
"sigma": 1,
"epsilon": 1,

"queuing":"slurm",
"queue_name":"medium",
"queue_template": "qu.tpl",
"queue_job_limit": 300,
"ppn": 1,
"scratch": "/scratch/cmartia",
"username": "cmartia",
"delete_intermediate_files": 0
}`

**Comment by cmartia** [comment]
I'll mark this solved for now. Feel free to reopen it if you still need help with the NN potentials.

**Comment by IliasChair** [comment]
Hello again,
Thanks to your help, I got it running. Besides the queuing section, I had to add these parameters to the *.json file for Kinbot to work:

```
"scan_method": "am1",
"scan_basis": "",
"method": "",
"basis": "",
"high_level_method": "",
"high_level_basis":"",
```
I assume these are left blank because we're using the NN PES and don't need to set them?

I still have a couple of quick questions:
To clarify we're aiming to apply the workflow from your [publication](DOI: 10.1021/acs.jpca.3c07872) to our system of interest using our own NN PES architecture.

The simulation for the above JSON is running fine, but I noticed Gaussian calculations are still happening in the background. While the Sella calculations seem to use the NN PES, there are still a lot of geometry optimizations and frequency calculations being done by Gaussian with the AM1 method. Is this expected? If yes, why so? Is there an option to use the NN PES for all calculations?

Thanks again for all your help and continued support - it's been really valuable!

Best 
Ilias
